### ISSUE TITLE
tf.cond should be evaluated lazily

### ISSUE THREAD

**Issue by 3rd3** [tensorflow/tensorflow#10800]
It should be possible to statically analyze the following graph such that the execution of the one branch is stopped early if the other branch is determined to be chosen. Am I missing some different function in TensorFlow that allows to do this? Perhaps using control dependencies?

In the following code the identity matrix is multiplied many times and a placeholder threshold determines whether the second multiplication or the last is chosen as output of the graph.

```
import tensorflow as tf
import numpy as np
import timeit

N = 2048

x = tf.placeholder(tf.float32, [N,N])
t = tf.placeholder(tf.float32)
net = x
stop = None
for i in range(300):
  net = tf.matmul(net, net)
  if i == 1:
    stop = net

net = tf.cond(tf.reduce_sum(stop) < t, lambda: stop * 0., lambda: net)

sess = tf.Session(config=tf.ConfigProto(log_device_placement=True))

r = None

start_time = timeit.default_timer()
for i in range(10):
  r = sess.run(net, feed_dict={x: np.eye(N), t: N})
print(r)
print(timeit.default_timer() - start_time)

start_time = timeit.default_timer()
for i in range(10):
  r = sess.run(net, feed_dict={x: np.eye(N), t: N + 1})
print(r)
print(timeit.default_timer() - start_time)
```

Both runs take roughly the same time even though the second run only depend on the first two matrix multiplications (at least if one accounts for the warm-up phase which give the first run a slight disadvantage).

I hope I am not missing something obvious. Thanks.

**Comment by aselle** [comment]
See #3287 #8168 and https://stackoverflow.com/questions/34401714/is-tensorflow-lazy for background info.

**Comment by albertz** [comment]
Only #3287 is related (the others don't use `tf.cond`). There you get a more detailed answer. Short answer: The creation of all the operations which should be executed lazily must be done inside the function call. In your case, you create them and inside the function call just refer to them.

This is not a bug what you see, you are using it wrong. Maybe the documentation in TF could be improved.

**Comment by 3rd3** [comment]
Thank you. I guess this solves my problem. Still, wouldn't it be possible to statically analyze the dependencies of a `tf.cond` nodes? I mean aren't such optimizations that the advantage of defining computations symbolically/declaratively in TensorFlow?

**Comment by albertz** [comment]
It's this way because TF was designed to be this way. It certainly could have been different, if dependencies of ops would be handled different in TF. But right now, when an op is going to be evaluated, all tensors which go into it must also be evaluated.

Note that it is quite interesting to check out the code of `tf.cond`. Internally, `tf.cond` is implemented via `switch`/`merge` and control flow contexts.

**Comment by yaroslavvb** [comment]
@3rd3 it's possible to do this by modifying the graph on the client level. Lazy evaluation is implemented using `switch/merge` statements, and `tf.cond` inserts those statements around the tensors created in `tf.cond` call. If you want to add laziness to tensors that have already been created, you can use `graph_editor` to modify the graph with the proper structure.

For instance, here's an example of modifying the graph to run variable initializer lazily -- https://gist.github.com/yaroslavvb/d67410e240369736fc4ba0267250ef27

**Comment by 3rd3** [comment]
@yaroslavvb Thanks, very interesting gist. Wouldn't it make sense to make these kinds of optimizations part of TensorFlow?

**Comment by yaroslavvb** [comment]
@3rd3

Note that reason for current semantics is that it matches what you would expect from other programming languages:

IE consider

```
1+2 if False else 2+3
```
vs

```
a=1+2
a if False else b
```

Like in your case with `tf.cond`, computation `a` is getting evaluated despite `if` being lazy. The way to make it truely lazy both in Python and TensorFlow case, is to move the actual computation inside `if`.

Changing existing `tf.cond` to work lazily on computation previously defined, would change the semantics. IE, suppose you have `tf.cond(a,b,c)`. Tensors `b` and `c` may have side-effects such as triggering variable assignments, and if you optimize it out, it'll change behavior/break existing TF scripts.

It might be useful to have something like `lazycond(a,b,c)` construct which has semantics you want. It could be done purely on Python level by analyzing graph and inserting appropriate switch/merge statements.

**Comment by 3rd3** [comment]
I am wondering why you use imperative semantics here when static analysis and optimization is exactly the strong suit of declarative programming.

It also seems that TensorFlow is already lazy in some regards because in the following case `y` is not computed since it is not queried.

    x = tf.no_op()
    y = tf.no_op()
    sess.run(x)

Query laziness, but no branching laziness? This seems to me like an odd mix of different paradigms.

**Comment by yaroslavvb** [comment]
@3rd3 indeed, this is a common pitfall. In retrospect I think it would be less confusion if your example usage was explicitly forbidden. You have a mix -- some Tensors are created inside tf.cond, while others are created outside of tf.cond. To ensure laziness, all computation must be defined by the lambda called by `tf.cond`, that enables the construct to insert proper merge/switch calls

**Comment by 3rd3** [comment]
But that would still not resolve the paradigm discrepancy between query laziness and branching non-laziness.

**Comment by yaroslavvb** [comment]
I would say this is Python API issue, rather than core TensorFlow issue,
since there's no "cond" in TF API, only Switch/Merge. Changing tf.cond to
be fully lazy will change behavior​ of existing programs, so it can't be
done for backward compatibility reasons. You could write your own version
of tf.cond which inserts Switch/Merge in a more intuitive way

On Jul 1, 2017 2:29 PM, "3rd3" <notifications@github.com> wrote:

But that would still not resolve the paradigm discrepancy between query
laziness and branching non-laziness.

—
You are receiving this because you were mentioned.

Reply to this email directly, view it on GitHub
<https://github.com/tensorflow/tensorflow/issues/10800#issuecomment-312429571>,
or mute the thread
